\documentclass{article}
\usepackage{parskip}
\usepackage[margin=.6in]{geometry}
\begin{document}
\section{Dynamic Multilevel Exit}
Many routines have non primary exits (dont go to the next statement after the call). All looping structures are stealthily gotos. Despite this we should only ever manually use gotos to simulare a labled break.

Usually labels are constants, but by saying \texttt{label L} we make a variable label. When we see a \texttt{goto L} we have to go into the stack frame to see which L we are going to since L is a variable (there are different versions of L based on what iteration we are on). We do this be unwinding the stack (popping off frames) until we find the right L (there can be an arbitrary number of hs, gs, and fs). We can accidentally transfer to a stack frame that no longer exists (baaaaaaad).

C has functions that allow us to do this (setjmp, longjmp, and jmpbuf).

\section{Traditional Approaches}
Return codes are a value returned that say if the routine was successful or if it is calling secondary/tertiary/etc results.
    \begin{itemize}
        \item programmers are lazy and dont check
        \item need a value that for sure wont be returned
    \end{itemize}

Status flags are global variables that we set based on things going sideways in the routine.
\begin{itemize}
    \item programmers are lazy
    \item these variables can get overwritten
\end{itemize}

Fixup routines is a routine passed in that attempts to fix problems in the routine so that no errors are thrown.
\begin{itemize}
    \item requires the fixup routine to get passed through a parent functions which may never use it.
\end{itemize}

When a routine goes bad it sets a flag errno to 22. Specifically we can call perror which checks the errno value and print out the message so that we can see this.


\subsection{Exception Handling}
Exceptions provide a more brutal break in the code. These are much harder for the programmer to ignore. Its pretty hard to simulate them though.

\subsection{Execution Envtoironment}
When we use jump commands in C destruction are not called. Its a much faster way to move around the stack, but it can cause massive leaks.

We run into some fun shenanigans when we have multiple stacks (like on threads). Usually at the end of a stack if an exception is not handled program aborts, but with this we can chuck that exception onto the next stack.

\section{Terminology}
\textbf{Execution} - anything that has a stack (where you can raise an excpetion)

\textbf{exception type} - what you throw

\textbf{exception} - follows along with excpetion type

\textbf{raise} - create an exception

\textbf{source execution} - creates exception

\textbf{faulting execution} - execution which changes to deal with exception

\textbf{local exception} - exception raised and handled by same exeuction

\textbf{non-local} - exeption gets passed to new execution

\textbf{concurrent exception} - source and faulting execution are concurrent

\textbf{propogation} - passes control to faulting after raise

\textbf{propogation mechanism} - rules for how control is passed (can be complicated, determined by language)

\textbf{handler} - routine that deals with exception

\textbf{rereaise} - handler cannot deal with exception so it reraises it to let it go back down the stack and hope someone else can deal with it

\textbf{guarded block} - block that can handle an excpetion

\textbf{unguarded block} - block that cannot handle exceptions

\textbf{termination} - if handler cannot recover everything is shit, do not come back here, just go die

\textbf{resumption} - opposite of resumption, things can be recovered if handled properly

\section{Static/Dynamic Call/Return}
Static part of program is one that can be determined by looking at the code without executing it. Things that can be determined only during or after execution are dynamic parts.

Say we have a call to function foo. We can statically determine which routine is going to be called. The same cannot be said for the return. Since foo can be called multiple times we don't necissarily know where foo will return to. The computer must store (dynamically) where the function is called from. Parts of the routine are static and parts are raised. Routines have a static call and a dynamic return (see table).

A call can be dynamic if we use pointers so that we don't actually know where the function has been called from since the point can be dereferenced anywhere and called.

\subsection{Static Propogation/Sequels}
When a sequel is called it does its shenanigans, but returns to the end of the block in which it was declared. This results in a static call (a sequel is just called normally) and a static return. Since the sequel can be only declared once, all calls to it will return to the same location (since its return location is based on its declaration location).

A weird example of this is a throw catch. When a catch is executed it transfers to the end of the block in which it is declared (always returns to the same location, and is called at the same location). Catches are just sequels
\end{document}
